eth/stagedsync/stage_polygon_sync.go

```diff
@@ -165,25 +171,214 @@ func SpawnPolygonSyncStage(
 		return err
 	}
 
-	if useExternalTx {
-		return nil
+	if !useExternalTx {
+		return tx.Commit()
+	}
+
+	return nil
+}
+
+func UnwindPolygonSyncStage(ctx context.Context, tx kv.RwTx, u *UnwindState, cfg PolygonSyncStageCfg) error {
+	u.UnwindPoint = max(u.UnwindPoint, cfg.blockReader.FrozenBlocks()) // protect from unwind behind files
+
+	useExternalTx := tx != nil
+	if !useExternalTx {
+		var err error
+		tx, err = cfg.db.BeginRw(ctx)
+		if err != nil {
+			return err
+		}
+
+		defer tx.Rollback()
+	}
+
+	// headers
+	unwindBlock := u.Reason.Block != nil
+	if err := rawdb.TruncateCanonicalHash(tx, u.UnwindPoint+1, unwindBlock); err != nil {
+		return err
+	}
+
+	canonicalHash, err := rawdb.ReadCanonicalHash(tx, u.UnwindPoint)
+	if err != nil {
+		return err
+	}
+
+	if err = rawdb.WriteHeadHeaderHash(tx, canonicalHash); err != nil {
+		return err
+	}
+
+	// bodies
+	if err = cfg.blockWriter.MakeBodiesNonCanonical(tx, u.UnwindPoint+1); err != nil {
+		return err
+	}
+
+	// heimdall
+	if err = UnwindHeimdall(tx, u, nil); err != nil {
+		return err
 	}
 
-	if err := tx.Commit(); err != nil {
+	if err = u.Done(tx); err != nil {
 		return err
 	}
 
+	if !useExternalTx {
+		return tx.Commit()
+	}
+
 	return nil
 }
 
-func UnwindPolygonSyncStage() error {
-	// TODO - headers, bodies (including txnums index), checkpoints, milestones, spans, state sync events
+func UnwindHeimdall(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if err := UnwindEvents(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindSpans(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindCheckpoints(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindMilestones(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func PrunePolygonSyncStage() error {
-	// TODO - headers, bodies (including txnums index), checkpoints, milestones, spans, state sync events
-	return nil
+func UnwindEvents(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if len(unwindTypes) > 0 && !slices.Contains(unwindTypes, "events") {
```

Comment:
I think this logic (of checking the unwind type) can be moved to UnwindHeimdall

Reply:
no it can't be because each function passes a different unwind type to the `Contains` function, e.g. `events`, `spans`, `milestones`, etc.

Reply:
I think that is better to be within the corresponding UnwindEvents/Spans/Milestones/etc funcs